### ISSUE TITLE
New Sonos developer API

### ISSUE THREAD

**Issue by ghcs27** [SoCo/SoCo#641]
I recently discovered that Sonos has released a new developer API at [developer.sonos.com](https://developer.sonos.com).
It is aimed at music service companies like Spotify who would like to include Sonos controls in their app.

Some aspects I have found out:
- Officially supported by Sonos
- Cloud-based communication over HTTP
- Requires the developer to register their integration with Sonos
- Requires the user to authorize the integration with their Sonos account (over OAuth)
- Contains only a subset of the features of the UPnP API - For instance, there's no way to play a music library item (only stream or cloud queue), set an alarm or access the player settings.

On the one hand, the official support is promising, but the required registration and limited feature set are major drawbacks. Still, it might help us resolve issues like #557. Sonos will also probably expand the API in the future (and might even replace the UPnP API with it?).

What do you think? Is it worth it to support this new API in our project?

**Comment by DPH** [comment]
This is good news. Sonos have been promising a proper API for some while. 
Our current UPnP works for most things but can't any longer work for music services. 
These are a big part of using Sonos. 
I'm very supportive of enhancing SoCo to use the Sonos API to access the Music Services.
Willing to help but will be very busy on other stuff for the next 2 months.

Regards Daivd

**Comment by relevitt** [comment]
My preference would be that SoCo doesn't support the new Sonos developer API. I see SoCo and the new API as alternative (and possibly complimentary) ways of interacting with Sonos speakers. There's nothing to stop someone with an app that uses SoCo from also using the new API. But I think SoCo would become unwieldy and complex if it tried to do both. I hope SoCo continues to focus on what it does (i.e. it's a library for a LAN based Sonos controller) and do it well.

I also don't see this new developer API replacing SoCo any time soon. For me, the requirement for cloud-based communication over HTTPS is a deal breaker. If my Internet connection goes down, I can play music from my local collection and control my Sonos system on the LAN using SoCo (or the Sonos app on my phone). Also, registration and authentication are a pain (and obviously a necessary consequence of the developer API being cloud-based), although I have those problems already with the Spotify API.

As to Music Service Authentication, I abandoned Sonos for searching and use the music service provider's API, being Spotify's in my case. The Spotify API is quicker, easier to use and has more features. 

Spotify Connect (now part of the Spotify API) can also be used to control Sonos speakers, but is too limited to replace SoCo as a controller. However, if one wants to add new items to a Spotify queue being played over Sonos via Spotify Connect (and lots of people now use Spotify Connect in this way), then one has to use the Spotify Connect API, because the queue exists only on a Spotify device, and not on the Sonos device. So, developers of Sonos controllers may find they want to use more than one API.

**Comment by stefankoegl** [comment]
I agree with @relevitt that trying to do both would risk making the library more complex, with limited benefit. In my opinion the project should keep its focus on the UPnP-based API, and do that well. A client library for the Cloud API could (should) be a separate project.
